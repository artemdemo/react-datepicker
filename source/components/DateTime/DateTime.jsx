import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import classnames from 'classnames';
import Calendar from './Calendar/Calendar';

import './DateTime.less';

class DatePicker extends React.Component {
    constructor(props) {
        super(props);
        const { value, utc } = this.props;
        const date = value;
        if (utc) {
            date.utc();
        }
        this.state = {
            showCalendar: false,
            inputValue: moment.isMoment(date) ? date.format(this.getFormat()) : '',
            date,
        };
        this.inputRef = null;
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.value !== this.props.value) {
            const date = nextProps.value;
            if (nextProps.utc) {
                date.utc();
            }
            this.setState({
                date,
            });
        }
    }

    setInputRef(ref) {
        this.inputRef = ref;
    }

    getFormat() {
        const { dateFormat, timeFormat } = this.props;
        return `${dateFormat} ${timeFormat}`;
    }

    inputFocusHandler() {
        this.setState({
            showCalendar: true,
        });
    }

    handleClickOutside(e) {
        if (e.target !== this.inputRef) {
            this.setState({
                showCalendar: false,
            });
        }
    }

    handleDateClick(newDate) {
        const { onChange } = this.props;
        this.setState({
            date: newDate,
            inputValue: newDate.format(this.getFormat()),
        });
        onChange && onChange(newDate);
    }

    handleInputChange(e) {
        const inputValue = e.target.value;
        const date = moment(e.target.value, this.getFormat());
        this.setState(Object.assign(
            { inputValue },
            date.isValid() ? { date } : null,
        ));

        const { onChange } = this.props;
        onChange && onChange(date.isValid() ? date : inputValue);
    }

    renderCalendar() {
        if (this.state.showCalendar) {
            const { timeFormat } = this.props;
            return (
                <div className='datetime-calendar-container'>
                    <Calendar
                        date={this.state.date}
                        onChange={this.handleDateClick.bind(this)}
                        timeFormat={timeFormat}
                        onClickOutside={this.handleClickOutside.bind(this)} />
                </div>
            );
        }
        return null;
    }

    render() {
        const { className } = this.props;
        const datetimeClass = classnames('datetime', className);
        return (
            <div className={datetimeClass}>
                <input
                    ref={this.setInputRef.bind(this)}
                    value={this.state.inputValue}
                    onChange={this.handleInputChange.bind(this)}
                    onFocus={this.inputFocusHandler.bind(this)}
                    className='datetime-input' />
                {this.renderCalendar()}
            </div>
        );
    }
}

DatePicker.propTypes = {
    value: PropTypes.instanceOf(moment),
    onChange: PropTypes.func,
    utc: PropTypes.bool,
    className: PropTypes.string,
    dateFormat: PropTypes.string,
    timeFormat: PropTypes.string,
};

DatePicker.defaultProps = {
    value: moment(),
    onChange: null,
    utc: false,
    className: '',
    dateFormat: 'YYYY-MM-DD',
    timeFormat: 'HH:mm:ss',
};

export default DatePicker;
