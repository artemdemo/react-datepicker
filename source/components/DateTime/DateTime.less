@import 'styles/general';
@import 'styles/animations';

.datetime {
    color: @text-color;
    user-select: none;
}

.datetime-calendar-container {
    position: absolute;
}
