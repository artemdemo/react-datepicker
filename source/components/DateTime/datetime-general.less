@hover-color: lightblue;
@text-color: black;

.datetime-table() {
    border-spacing: 0;
    border-collapse: collapse;
    table-layout: fixed;
    width: 100%;
}
