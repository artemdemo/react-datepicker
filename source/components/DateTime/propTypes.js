/* eslint-disable max-len */
import moment from 'moment';

const isRequiredFactory = (propFunc) => {
    propFunc.isRequired = (props, propName, componentName) => {
        if (props[propName] === null) {
            return new Error(
                `The prop \`${propName}\` is marked as required (in \`${componentName}\`), but its value is \`null\``,
            );
        } else if (props[propName] === undefined) {
            return new Error(
                `The prop \`${propName}\` is marked as required (in \`${componentName}\`), but its value is \`undefined\``,
            );
        }
        return propFunc(props, propName, componentName);
    };
    return propFunc;
};

export const propIsMoment = isRequiredFactory((props, propName, componentName) => {
    if (!moment.isMoment(props[propName])) {
        return new Error(
            `Invalid prop \`${propName}\` supplied to \`${componentName}\`. Validation failed.`,
        );
    }
    return undefined;
});
