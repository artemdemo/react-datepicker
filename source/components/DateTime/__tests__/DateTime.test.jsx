import React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import DateTime from '../DateTime';

describe('<DateTime />', () => {
    it('simple render', () => {
        const tree = renderer.create(
            <DateTime
                defaultValue={moment(1502459094000)}
                utc />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
