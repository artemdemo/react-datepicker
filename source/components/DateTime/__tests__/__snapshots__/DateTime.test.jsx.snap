// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateTime /> inputProps 1`] = `
<div
  className="datetime"
>
  <input
    className="datetime-input some-class"
    disabled={true}
    name="Some name"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Some placeholder"
    required={true}
    value="2017-08-11 13:44:54"
  />
</div>
`;

exports[`<DateTime /> simple render 1`] = `
<div
  className="datetime"
>
  <input
    className="datetime-input"
    disabled={false}
    name={undefined}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    required={false}
    value="2017-08-11 13:44:54"
  />
</div>
`;
